### PR TITLE
[DO NOT MERGE] Always prespecialize generic metadata.

### PR DESCRIPTION
**Pull request by nate-chandler**
Leaving open to benchmark and test source compatibility of turning on metadata prespecialization always, independent of target version.

### CODE REVIEW

**Comment by nate-chandler** [comment]
@swift-ci please smoke benchmark

File: lib/IRGen/GenDecl.cpp [hunk 1]
```diff
@@ -1132,6 +1133,13 @@ void IRGenerator::emitLazyDefinitions() {
       CurrentIGMPtr IGM = getGenModule(type->getDeclContext());
       emitLazyTypeMetadata(*IGM.get(), type);
     }
+    while (!LazySpecializedTypeMetadataRecords.empty()) {
+      CanBoundGenericType type =
```
Comment (slavapestov) [inline lib/IRGen/GenDecl.cpp]: What about pre-specializing types such as `Foo<T>.Bar`?

File: lib/IRGen/GenDecl.cpp [hunk 2]
```diff
@@ -3595,8 +3609,10 @@ llvm::GlobalValue *IRGenModule::defineTypeMetadata(CanType concreteType,
   if (nominal)
     addRuntimeResolvableType(nominal);
 
-  // Don't define the alias for foreign type metadata, since it's not ABI.
-  if (nominal && requiresForeignTypeMetadata(nominal))
+  // Don't define the alias for foreign type metadata or prespecialized generic
+  // metadata, since neither is ABI.
+  if ((nominal && requiresForeignTypeMetadata(nominal)) ||
+      concreteType->is<BoundGenericType>())
```
Comment (slavapestov) [inline lib/IRGen/GenDecl.cpp]: Another suspicious use of BoundGenericType.

**Comment by nate-chandler** [comment]
Build failure:

```
18:29:59 FAILED: lib/swift/macosx/x86_64/libswiftRemoteMirror.dylib 
18:29:59 : && /Users/buildnode/jenkins/workspace/swift-PR-osx-smoke-perf/Ninja-Release/llvm-macosx-x86_64/./bin/clang++ -Wno-unknown-warning-option -Werror=unguarded-availability-new -fno-stack-protector -fPIC -fvisibility-inlines-hidden -Werror=date-time -Werror=unguarded-availability-new -std=c++14 -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -Wimplicit-fallthrough -Wcovered-switch-default -Wno-class-memaccess -Wno-noexcept-type -Wnon-virtual-dtor -Wdelete-non-virtual-dtor -Wstring-conversion -fdiagnostics-color -Werror=switch -Wdocumentation -Wimplicit-fallthrough -Wunreachable-code -Woverloaded-virtual -DOBJC_OLD_DISPATCH_PROTOTYPES=0 -O3 -DNDEBUG -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk -dynamiclib -Wl,-headerpad_max_install_names -dynamiclib\ -Wl,-headerpad_max_install_names -Xlinker -compatibility_version -Xlinker 1 -target x86_64-apple-macosx10.9 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk -arch x86_64 -F /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/../../../Developer/Library/Frameworks -mmacosx-version-min=10.9 -Wl,-dead_strip  "-L/Users/buildnode/jenkins/workspace/swift-PR-osx-smoke-perf/Ninja-Release/swift-macosx-x86_64/./lib/swift/macosx/x86_64" "-L/Users/buildnode/jenkins/workspace/swift-PR-osx-smoke-perf/Ninja-Release/swift-macosx-x86_64/./bin/../lib/swift/macosx/x86_64" "-L/Users/buildnode/jenkins/workspace/swift-PR-osx-smoke-perf/Ninja-Release/swift-macosx-x86_64/./bin/../lib/swift/macosx" -o lib/swift/macosx/x86_64/libswiftRemoteMirror.dylib -install_name @rpath/libswiftRemoteMirror.dylib stdlib/public/SwiftRemoteMirror/CMakeFiles/swiftRemoteMirror-macosx-x86_64.dir/SwiftRemoteMirror.cpp.o -L/Users/buildnode/jenkins/workspace/swift-PR-osx-smoke-perf/Ninja-Release/llvm-macosx-x86_64/./lib lib/swift/macosx/x86_64/libswiftReflection.a && :
18:29:59 Undefined symbols for architecture x86_64:
18:29:59   "swift::Demangle::Node::dump()", referenced from:
18:29:59       (anonymous namespace)::Remangler::mangle(swift::Demangle::Node*) in libswiftReflection.a(Remangler.cpp.o)
18:29:59 ld: symbol(s) not found for architecture x86_64
```

**Comment by nate-chandler** [comment]
@swift-ci please clean smoke benchmark

**Comment by nate-chandler** [comment]
@swift-ci please clean smoke test

File: lib/IRGen/GenDecl.cpp (hunk 2 above)
Comment (nate-chandler) [reply lib/IRGen/GenDecl.cpp]: Yeah, that limitation is everywhere at the moment.  Would it be reasonable to start out only prespecializing these cases and expand from here?

**Comment by nate-chandler** [comment]
@swift-ci please clean smoke test

File: lib/IRGen/GenDecl.cpp (hunk 1 above)
Comment (nate-chandler) [reply lib/IRGen/GenDecl.cpp]: Should work now.

**Comment by nate-chandler** [comment]
@swift-ci please smoke test

**Comment by nate-chandler** [comment]
@swift-ci please smoke test

**Comment by nate-chandler** [comment]
@swift-ci please benchmark

**Comment by nate-chandler** [comment]
@swift-ci please smoke test

**Comment by nate-chandler** [comment]
@swift-ci please smoke test

**Comment by nate-chandler** [comment]
@swift-ci please smoke test

**Comment by nate-chandler** [comment]
@swift-ci please benchmark

**Comment by nate-chandler** [comment]
@swift-ci please benchmark

**Comment by nate-chandler** [comment]
@swift-ci please benchmark

**Comment by nate-chandler** [comment]
@swift-ci Please Test Source Compatibility

**Comment by nate-chandler** [comment]
@swift-ci please benchmark

**Comment by nate-chandler** [comment]
@swift-ci please benchmark

**Comment by nate-chandler** [comment]
@swift-ci please benchmark

**Comment by nate-chandler** [comment]
@swift-ci please test compiler performance

**Comment by nate-chandler** [comment]
Release:

```
LLVM.NumLLVMBytesOutput | 1,923,558,944 | 1,920,178,772 | -3,380,172 | -0.18%
```

Debug:

```
LLVM.NumLLVMBytesOutput | 1,800,361,224 | 1,801,765,732 | 1,404,508 | 0.08%
```

Source:

https://ci.swift.org/job/swift-PR-compiler-performance-macOS/571/artifact/

**Comment by nate-chandler** [comment]
@swift-ci please test source compatibility

**Comment by nate-chandler** [comment]
@swift-ci please benchmark

**Comment by nate-chandler** [comment]
@swift-ci please test compiler performance

**Comment by nate-chandler** [comment]
Release:
```
LLVM.NumLLVMBytesOutput | 1,923,705,960 | 1,921,602,736 | -2,103,224 | -0.11%
```

Debug:
```
LLVM.NumLLVMBytesOutput | 1,800,469,268 | 1,802,125,694 | 1,656,426 | 0.09%
```

Source:

https://ci.swift.org/job/swift-PR-compiler-performance-macOS/576/artifact/comment.md

**Comment by nate-chandler** [comment]
@swift-ci please test source compatibility

**Comment by nate-chandler** [comment]
@swift-ci please test source compatibility

**Comment by nate-chandler** [comment]
@swift-ci please test source compatibility

**Comment by nate-chandler** [comment]
@swift-ci please test source compatibility

**Comment by nate-chandler** [comment]
@swift-ci please benchmark

**Comment by nate-chandler** [comment]
@swift-ci please test source compatibility

**Comment by nate-chandler** [comment]
@swift-ci please smoke benchmark

**Comment by nate-chandler** [comment]
@swift-ci please benchmark

**Comment by nate-chandler** [comment]
@swift-ci please test

**Comment by nate-chandler** [comment]
@swift-ci please test source compatibility

**Comment by nate-chandler** [comment]
@swift-ci please benchmark

**Comment by nate-chandler** [comment]
@swift-ci please test source compatibility

**Comment by nate-chandler** [comment]
@swift-ci please test source compatibility

**Comment by nate-chandler** [comment]
@swift-ci please benchmark

**Comment by nate-chandler** [comment]
@swift-ci please test source compatibility

**Comment by nate-chandler** [comment]
@swift-ci please test

**Comment by nate-chandler** [comment]
@swift-ci please clean test linux platform

**Comment by futurejones** [comment]
Can one of the admins verify this patch?

**Pull request closed without merging**